dg-commons/src/main/code/Graph/Graph.scala

```diff
@@ -0,0 +1,210 @@
+
+    newNode
+  }
+
+  def getNodeByDisplayId(displayableDataId: String): Option[Node[T]] = {
+    allNodes.find(node => node.data.displayableDataId.equals(displayableDataId))
+  }
+
+  /**
+   * Writes this graph to a GraphViz "DOT" file -- there are various clients to read this kind of file, save to image, etc.
+   * @param filepathToCreate
+   * @param isSimplified
+   * @param alsoWriteAsPng
+   */
+  def writeDotFile(filepathToCreate: String, isSimplified: Boolean = false, alsoWriteAsPng: Boolean = true) = {
+    val writer = new FileOutputStream(filepathToCreate)
+    try {
+      writeDotFileToOpenStream(writer, isSimplified = isSimplified)
+    }
+    finally {
+      writer.close()
+    }
+
+    if (alsoWriteAsPng) {
+      DotHelper.writeDotFileAsPng(filepathToCreate)
+    }
+  }
+
+  /**
+   * Writes this graph to a GraphViz "DOT" file stream -- there are various clients to read this kind of file, save to image, etc.
+   * @param out
+   * @param isSimplified
+   */
+  def writeDotFileToOpenStream(out: OutputStream, isSimplified: Boolean = false): Unit = {
+    // Write any graph-specific attributes
+    out.write(s"digraph Graph_${graphId} {\r\n".getBytes("UTF-8"))
+
+    // Write all labels and edges
+    allNodes.foreach(_.writeDotFormatGraphVisualizationOfNodeToOpenStream(out, isSimplified = isSimplified))
+
+    // End graph
+    out.write("}".getBytes("UTF-8"))
+  }
+
+  /**
+   * Create a copy of this graph
+   * @return
+   */
+  def deepCopy: Graph[T] = {
+    //this.deepClone // Has performance issues (uses reflection)... so let's implement it ourselves instead.
+    val copiedGraph = new Graph[T](initialNodeValue = Some(allNodes(0).data), isEdgeLinkTrackingOn = false, _graphId = _graphId, appendSharedDisplayIdsWithNumericalSuffix = false)
+    copiedGraph.edgeLinkTrackingDescriptions = edgeLinkTrackingDescriptions
+    copiedGraph.customSeed = customSeed
+    copiedGraph.customGlobalSeed = customGlobalSeed
+    copiedGraph.nodeIdCounters = nodeIdCounters
+    //copiedGraph.customBooleanAttributes = customBooleanAttributes
+    allNodes.tail.foreach(node => copiedGraph.addNewRootNode(node.data))
+    allNodes.foreach(node => {
+      val nodeInNewGraph = copiedGraph.allNodes(node.nodeIndexInContainingGraph)
+      node.parents.foreach(parent => {
+        val parentInNewGraph = copiedGraph.allNodes(parent.nodeIndexInContainingGraph)
+        if (!nodeInNewGraph.parents.contains(parentInNewGraph)) {
+          nodeInNewGraph.addLinkToExistingParent(parentInNewGraph)
+        }
+      })
+    })
+
+    copiedGraph.isEdgeLinkTrackingOn = isEdgeLinkTrackingOn
+    copiedGraph.appendSharedDisplayIdsWithNumericalSuffix = appendSharedDisplayIdsWithNumericalSuffix
+    copiedGraph
+  }
+
+  /*
+  /**
+   * Used for testing graph isomorphism.
+   * @return
+   */
+  def getStructuralMD5: String = {
+    // TODO: Implement. This is a graph isomorphism problem, which is not yet known if possible in polynomial time.
```

Comment:
That's ok. You can leave the todo but please add tickets so that we don't
forget what that todo was for and reference yourself in the tickets. (treat
other todos in a similar way)

On Tue, Jul 21, 2015 at 11:21 PM, SamerAdra <notifications@github.com>
wrote:

> In dg-commons/src/main/code/Graph/Graph.scala
> <https://github.com/FINRAOS/DataGenerator/pull/222#discussion_r35178359>:
>
> > +        }
> > +      })
> > +    })
> > +
> > +    copiedGraph.isEdgeLinkTrackingOn = isEdgeLinkTrackingOn
> > +    copiedGraph.appendSharedDisplayIdsWithNumericalSuffix = appendSharedDisplayIdsWithNumericalSuffix
> > +    copiedGraph
> > +  }
> > +
> > +  /*
> > +  /**
> > +   * Used for testing graph isomorphism.
> > +   * @return
> > +   */
> > +  def getStructuralMD5: String = {
> > +    // TODO: Implement. This is a graph isomorphism problem, which is not yet known if possible in polynomial time.
>
> It is a placeholder for a future feature -- if I had such a method it
> would allow me to change the all-structures generation to prevent
> isomorphic graphs from being generated (all combinations instead of all
> permutations).
> … <#14eb3c7ac0757a42_>
> On Tuesday, July 21, 2015, Mohamed Ibrahim <notifications@github.com>
> wrote: In dg-commons/src/main/code/Graph/Graph.scala <
> https://github.com/FINRAOS/DataGenerator/pull/222#discussion_r35176694>:
> > + } > + }) > + }) > + > + copiedGraph.isEdgeLinkTrackingOn =
> isEdgeLinkTrackingOn > +
> copiedGraph.appendSharedDisplayIdsWithNumericalSuffix =
> appendSharedDisplayIdsWithNumericalSuffix > + copiedGraph > + } > + > + /*
> > + /** > + * Used for testing graph isomorphism. > + * @return
> <https://github.com/return> > + */ > + def getStructuralMD5: String = { >
> + // TODO: Implement. This is a graph isomorphism problem, which is not yet
> known if possible in polynomial time. Is that an incomplete code ? — Reply
> to this email directly or view it on GitHub <
> https://github.com/FINRAOS/DataGenerator/pull/222/files#r35176694>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FINRAOS/DataGenerator/pull/222/files#r35178359>.
>